### PR TITLE
[OpaqueValues] Handle PartialApplyInsts and @out tuples.

### PR DESCRIPTION
**Pull request by nate-chandler**
AddressLowering handles PartialApplyInsts as it handles other applies.

SILGen handles @out tuples by flattening the values when passed to the epilog block and by reconstructing the tuple in the epilog before it is returned.

Individual commit messages provide additional details.

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
The source compat failures are both in the type-checker and occurring with other PR requested runs.  Merging.

File: lib/SILGen/SILGenEpilog.cpp [hunk 1]
```diff
@@ -61,15 +84,84 @@ void SILGenFunction::prepareCoroutineUnwindEpilog(CleanupLocation cleanupLoc) {
...
+/// formal result type, except in the case of @out tuples where it must be
+/// preserved.
 static SILValue buildReturnValue(SILGenFunction &SGF, SILLocation loc,
                                  ArrayRef<SILValue> directResults) {
   if (directResults.size() == 1)
     return directResults[0];
 
+  auto fnConv = SGF.F.getConventions();
+  if (!fnConv.useLoweredAddresses()) {
+    // In opaque-values code, nested @out enums are not flattened.  Reconstruct
```
Comment (atrick) [inline lib/SILGen/SILGenEpilog.cpp:135]: This says "nested @out enums". Is it specific to tuples?

I'm still trying to convince myself that it is sufficient for this `buildReturnValue` logic to really be tuple specific, and specific to the function convention as opposed to the reabstraction pattern. Why is this only an issue for closures being passed to generic function types?

We should have @jckarter sign off on this approach

**Review by atrick — COMMENTED** [review]
This code all looks good. I'm still trying to understand the conditions under which tuples are expanded.

File: lib/SILGen/SILGenEpilog.cpp (hunk 1 above)
Comment (nate-chandler) [reply lib/SILGen/SILGenEpilog.cpp:135]: Yes, this is specific to tuples.  The flattening only occurs for tuples.

**Pull request merged**